cli/src/commands/rebase.rs

```diff
@@ -348,91 +351,107 @@ fn rebase_descendants_transaction(
-            rewriter.replace_parent(to_rebase_commit.id(), to_rebase_commit.parent_ids());
+            if target_roots.contains(&old_commit_id) {
+                rewriter.set_new_parents(new_parents.clone());
+            }
             if rewriter.parents_changed() {
                 let builder = rewriter.rebase(settings)?;
                 let commit = builder.write()?;
-                rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+                new_rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+            } else {
+                // Only include commit in skipped commit if it wasn't previously rewritten.
+                if target_commits
+                    .iter()
+                    .any(|commit| commit.id() == rewriter.old_commit().id())
+                {
+                    skipped_commits.push(rewriter.old_commit().clone());
+                }
             }
             Ok(())
         },
     )?;
 
-    let num_rebased_descendants = rebased_commit_ids.len();
-
-    // We now update `new_parents` to account for the rebase of all of
-    // `to_rebase_commit`'s descendants. Even if some of the original `new_parents`
-    // were descendants of `to_rebase_commit`, this will no longer be the case after
-    // the update.
-    let new_parents = new_parents
-        .iter()
-        .map(|new_parent| {
-            rebased_commit_ids
-                .get(new_parent.id())
-                .unwrap_or(new_parent.id())
-                .clone()
-        })
-        .collect();
-
-    let tx_description = format!("rebase commit {}", to_rebase_commit.id().hex());
-    // Finally, it's safe to rebase `to_rebase_commit`. We can skip rebasing if it
-    // is already a child of `new_parents`. Otherwise, at this point, it should
-    // no longer have any children; they have all been rebased and the originals
-    // have been abandoned.
-    let skipped_commit_rebase = if to_rebase_commit.parent_ids() == new_parents {
-        if let Some(mut formatter) = ui.status_formatter() {
-            write!(formatter, "Skipping rebase of commit ")?;
-            tx.write_commit_summary(formatter.as_mut(), &to_rebase_commit)?;
-            writeln!(formatter)?;
-        }
-        true
-    } else {
-        rebase_commit(settings, tx.mut_repo(), to_rebase_commit, new_parents)?;
-        debug_assert_eq!(tx.mut_repo().rebase_descendants(settings)?, 0);
-        false
-    };
-
-    if num_rebased_descendants > 0 {
-        if skipped_commit_rebase {
-            writeln!(
-                ui.status(),
-                "Rebased {num_rebased_descendants} descendant commits onto parent of commit"
+    if let Some(mut fmt) = ui.status_formatter() {
+        if !skipped_commits.is_empty() {
+            log_skipped_rebase_commits_message(
+                fmt.as_mut(),
+                tx.base_workspace_helper(),
+                &skipped_commits.iter().collect_vec(),
```

Comment:
I think it's required, since `log_skipped_rebase_commits_message` accepts `&[&Commit]` and not an iterator. Since the number of skipped commits is also being checked, not sure if converting to an iterator is the best choice also.

Reply:
`log_skipped_rebase_commits_message()` could be changed to accept `ExactSizeIterator`, but I don't have strong opinion.

Reply:
Oh, I didn't know there was an `ExactSizeIterator`, I can to use that instead.

Reply:
Ok, updated to use `ExactSizeIterator`.